imagetagger/imagetagger/images/models.py

```diff
@@ -68,10 +84,61 @@ class Meta:
         default=None
     )
     pinned_by = models.ManyToManyField(settings.AUTH_USER_MODEL, related_name='pinned_sets')
+    zip_finished = models.BooleanField(default=True)
 
     def root_path(self):
         return os.path.join(settings.IMAGE_PATH, self.path)
 
+    def zip_file_exists(self):
+        return os.path.isfile(self.zip_path())
+
+    def zip_path(self):
+        return os.path.join(settings.IMAGE_PATH, self.relative_zip_path())
+
+    def relative_zip_path(self):
+        return os.path.join(self.path, self.name) + '.zip'
+
+    def _update_zip_thread(self):
+        self.zip_finished = False
+        self.save()
+
+        # The python standard library module 'zipfile' is used for
+        # the zipping process. It allows adding of files with a
+        # different file name in the zip file which is crucial because
+        # the internal store process adds random strings to the
+        # filename that should not appear in the zip archive.
+        # Unfortunately, the module does not support deletion of
+        # single files in the archive. Therefore, the 'zip' command
+        # line tool is called as a subprocess.
+
+        with zipfile.ZipFile(self.zip_path(), 'a') as f:
+            contents = f.namelist()
+            # Add missing image to the archive
+            for image in self.images.all():
+                if image.name in contents:
+                    contents.remove(image.name)
+                else:
+                    f.write(image.path(), image.name)
+
+        # Remove deleted images
+        if len(contents) > 0:
+            command = ['zip', '-d', self.zip_path()] + contents
+            subprocess.run(command)
+
+        self.zip_finished = True
```

Comment:
This has to be updated in case of exceptions in the process, but raise the exception again to not shadow it